File: frontend/src/lib/sitemap.js
import { productsAPI, categoriesAPI } from '../lib/api';
import { SITE_CONFIG } from '../components/seo/seoConfig';

const changeFrequency = {
  homepage: 'daily',
  products: 'daily',
  product: 'weekly',
  category: 'weekly',
  static: 'monthly',
};

const priority = {
  homepage: '1.0',
  products: '0.9',
  product: '0.8',
  category: '0.7',
  static: '0.5',
};

const formatDate = (date) => {
  return new Date(date).toISOString().split('T')[0];
};

export const generateSitemap = async () => {
  const urls = [];
  const today = formatDate(new Date());

  urls.push({
    loc: `${SITE_CONFIG.url}/`,
    lastmod: today,
    changefreq: changeFrequency.homepage,
    priority: priority.homepage,
  });

  urls.push({
    loc: `${SITE_CONFIG.url}/products`,
    lastmod: today,
    changefreq: changeFrequency.products,
    priority: priority.products,
  });

  try {
    const [categoriesRes, productsRes] = await Promise.all([
      categoriesAPI.list(true),
      productsAPI.list({ page: 1, page_size: 1000 }),
    ]);

    const categories = categoriesRes.data || [];
    categories.forEach((category) => {
      urls.push({
        loc: `${SITE_CONFIG.url}/products?category=${category.slug}`,
        lastmod: today,
        changefreq: changeFrequency.category,
        priority: priority.category,
      });
    });

    const products = productsRes.data?.items || [];
    products.forEach((product) => {
      urls.push({
        loc: `${SITE_CONFIG.url}/products/${product.slug}`,
        lastmod: formatDate(product.updated_at || product.created_at),
        changefreq: changeFrequency.product,
        priority: priority.product,
      });
    });
  } catch (error) {
    console.error('Error fetching data for sitemap:', error);
  }

  return urls;
};

export const generateSitemapXml = async () => {
  const urls = await generateSitemap();

  const urlsXml = urls
    .map(
      (url) => `
  <url>
    <loc>${url.loc}</loc>
    <lastmod>${url.lastmod}</lastmod>
    <changefreq>${url.changefreq}</changefreq>
    <priority>${url.priority}</priority>
  </url>`
    )
    .join('');

  return `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
${urlsXml}
</urlset>`;
};

File: backend/routers/sitemap.py
"""Sitemap router for SEO."""
from fastapi import APIRouter, Response, Depends
from sqlalchemy.orm import Session
from datetime import datetime

from config.database import SessionLocal
from config.settings import settings

router = APIRouter(tags=["SEO"])

FRONTEND_URL = settings.FRONTEND_URL if hasattr(settings, 'FRONTEND_URL') and settings.FRONTEND_URL else "https://yourdomain.com"


def get_db():
    """Database session dependency."""
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


def format_date(dt):
    """Format datetime for sitemap."""
    if dt is None:
        return datetime.now().strftime("%Y-%m-%d")
    if isinstance(dt, str):
        return dt[:10] if len(dt) >= 10 else datetime.now().strftime("%Y-%m-%d")
    return dt.strftime("%Y-%m-%d")


@router.get("/sitemap.xml")
async def get_sitemap(db: Session = Depends(get_db)):
    """Generate XML sitemap for search engines."""
    
    from models.models import Product, Category
    
    urls = []
    today = datetime.now().strftime("%Y-%m-%d")
    
    urls.append(f"""  <url>
    <loc>{FRONTEND_URL}/</loc>
    <lastmod>{today}</lastmod>
    <changefreq>daily</changefreq>
    <priority>1.0</priority>
  </url>""")
    
    urls.append(f"""  <url>
    <loc>{FRONTEND_URL}/products</loc>
    <lastmod>{today}</lastmod>
    <changefreq>daily</changefreq>
    <priority>0.9</priority>
  </url>""")
    
    categories = db.query(Category).filter(
        Category.is_active == True,
        Category.is_deleted == False
    ).all()
    
    for category in categories:
        urls.append(f"""  <url>
    <loc>{FRONTEND_URL}/products?category={category.slug}</loc>
    <lastmod>{format_date(category.updated_at)}</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.7</priority>
  </url>""")
    
    products = db.query(Product).filter(
        Product.is_active == True,
        Product.is_deleted == False
    ).all()
    
    for product in products:
        urls.append(f"""  <url>
    <loc>{FRONTEND_URL}/products/{product.slug}</loc>
    <lastmod>{format_date(product.updated_at)}</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.8</priority>
  </url>""")
    
    sitemap_xml = f"""<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
{chr(10).join(urls)}
</urlset>"""
    
    return Response(
        content=sitemap_xml,
        media_type="application/xml"
    )


@router.get("/robots.txt")
async def get_robots():
    """Get robots.txt content."""
    robots_txt = f"""User-agent: *
Allow: /
Disallow: /admin/
Disallow: /checkout
Disallow: /orders
Disallow: /wishlist
Disallow: /cart
Disallow: /api/admin/
Disallow: /api/cart/

User-agent: Googlebot
Allow: /

Sitemap: {FRONTEND_URL}/api/sitemap.xml
"""
    return Response(
        content=robots_txt,
        media_type="text/plain"
    )
